The KubernetesPodOperator Comprehensive Guide (#15270)
mistercrunch/Airflow: docs/apache-airflow-providers-cncf-kubernetes/operators.rst

.. _howto/operator:KubernetesPodOperator:

KubernetesPodOperator - The Comprehensive Guide
===============================================

The :class:`~airflow.providers.cncf.kubernetes.operators.kubernetes_pod.KubernetesPodOperator` allows

> you to create and run Pods on a Kubernetes cluster. The task wrapped in the KubernetesPodOperator is then executed in
> these pods.

Reviewer: ```suggestion
you to create and run a Pod on a Kubernetes cluster as a task.
```

Maybe?

  :ref:`GKEStartPodOperator <howto/operator:GKEStartPodOperator>` operator as it
  simplifies the Kubernetes authorization process.

How does this operator work?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The :class:`~airflow.providers.cncf.kubernetes.operators.kubernetes_pod.KubernetesPodOperator` uses the
Kubernetes API to launch a pod in a Kubernetes cluster. By supplying an
image URL and a command with optional arguments, the operator uses the Kube Python Client to generate a Kubernetes API
request that dynamically launches those individual pods.

> Under the hood, :class:`~airflow.providers.cncf.kubernetes.hooks.kubernetes.KubernetesHook` creates the connection to
> the Kubernetes API server.

Reviewer: ```suggestion
```

I think this is redundant with the beginning of the this paragraph.

Essentially, KubernetesPodOperator packages all the supplied parameters into a request object which is then shipped off
to Kubernetes API Server so that the pod to execute your task is created. Whenever a task is triggered, a new worker pod
is spun up to execute that task. And once the task is completed, by default the worker pod is deleted

> and the resources reclaimed.

Reviewer: I'd be careful with terminology here. "Worker pods" are a KubeExecutor thing, not KPO.

Users can specify a kubeconfig file using the ``config_file`` parameter, otherwise the operator will default
to ``~/.kube/config``.

How does the KubernetesPodOperator differ from the KubernetesExecutor
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. note::
  The :doc:`Kubernetes executor <apache-airflow:executor/kubernetes>` is **not** required to use this operator.

What problems does KubernetesPodOperator solve?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

* The :class:`~airflow.providers.cncf.kubernetes.operators.kubernetes_pod.KubernetesPodOperator` enables task-level
  resource configuration and is optimal for custom Python dependencies that are not available through the
  public PyPI repository.

* It allows users to supply a template YAML file using the ``pod_template_file`` parameter.

* It allows isolation of deployments, configuration reuse, delegation and better management of secrets.

> * Ultimately, it allows Airflow to act a job orchestrator - no matter the language those jobs are written in.

Reviewer: I think a nice succinct way to think of it is: 'Easy way to run any image on Kubernetes as a task', and I feel like this should be first! (related to the orchestrator and any language points imo)

    :start-after: [START howto_operator_k8s_cluster_resources]
    :end-before: [END howto_operator_k8s_cluster_resources]

How to use KubernetesPodOperator with YAML file/JSON spec?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

>  * WIP

Reviewer 2: Does these `WIP`s mean this PR is still a draft at this moment? If so, let's convert it to draft.

Author: Yes. It is more of a draft now - lots more write-ups will be added.
